lib/ddq-backend-mysql.js

```diff
@@ -0,0 +1,386 @@
+                [
+                    ddqBackendInstance.config.table,
+                    ddqBackendInstance.config.topics
+                ],
+                callback
+            );
+
+            ddqBackendInstance.poller = timers.setTimeout(findData, ddqBackendInstance.config.pollDelayMs);
+        }
+
+        ddqBackendInstance.poller = timers.setTimeout(findData, ddqBackendInstance.config.pollDelayMs);
+    }
+
+
+    /**
+     * Updates a specific record so that it can be requeued.
+     *
+     * @param {Object} ddqBackendInstance
+     * @param {string} recordId
+     * @param {Function} callback
+     */
+    function requeue(ddqBackendInstance, recordId, callback) {
+        ddqBackendInstance.connection.query(`UPDATE ??
+            SET owner = null, isProcessing = false, requeued = true
+            WHERE hash = ?;`,
+            [
+                ddqBackendInstance.config.table,
+                recordId
+            ],
+            callback
+        );
+    }
+
+
+    /**
+     * Deletes a specific record from the database. If unsuccessful, this
+     * will call the command to requeue the record.
+     *
+     * @param {Object} ddqBackendInstance
+     * @param {string} recordId
+     */
+    function remove(ddqBackendInstance, recordId) {
+        ddqBackendInstance.connection.query(`DELETE FROM ??
+            WHERE hash = ?
+            AND requeued = false;`,
+            [
+                ddqBackendInstance.config.table,
+                recordId
+            ],
+            (err) => {
+                if (err) {
+                    console.error("Error deleting record with id:", recordId);
+                    console.error("Attempting requeue");
+                    requeue(ddqBackendInstance, recordId, (requeueErr) => {
+                        ddqBackendInstance.emit(EMIT_ERR, requeueErr);
+                    });
+                }
+            }
+        );
+    }
+
+
+    /**
+     * Ensures that heartbeatCleanup is ready to run according to the
+     * heartbeatCleanupDelay.
+     *
+     * @param {Object} ddqBackendInstance
+     */
+    function restore(ddqBackendInstance) {
+        /**
+         * Queries for jobs that have stopped, and primes them to be cleaned up.
+         */
+        function heartbeatCleanup() {
+            ddqBackendInstance.connection.query(`UPDATE ??
+                SET isProcessing = false, owner = null, requeued = false
+                WHERE DATEDIFF(NOW(), SEC_TO_TIME(?)) > heartbeatDate
+                AND isProcessing = true;`,
+                [
+                    ddqBackendInstance.config.table,
+                    ddqBackendInstance.config.heartbeatLifetimeSeconds
+                ],
+                (err) => {
+                    if (err) {
+                        ddqBackendInstance.emit(EMIT_ERR, err);
+                    }
+                }
+            );
+
+            ddqBackendInstance.restorer = timers.setTimeout(heartbeatCleanup,
```

Comment:
No newline at the end.

This reset of the restorer should happen when the query finishes, otherwise this has the same problem as setInterval.  See comments about poller.